### PR TITLE
drivers: add ms5611 barometer driver

### PR DESCRIPTION
**Pull request by karold2412**
This commit introduces ms5611 barometer driver. After applying this
commit user will be able to use ms5611 barometer in zephyr and set its
parameters like oversampling ratio via attr_set sensor API function.

It has been tested alongside lps25h barometer and barometer inside mobile
phone. Results were following:
                    MS5611:    LPS25H:     mobile phone:
ground level:       100,86kPa  101,88kPa   100,95kPa
72cm above ground:  100,85kPa  101,87kPa   100,94kPa
189cm above ground: 100,84kPa  101,85kPa   100,93kPa

Since ms5611 has a temperature compensation (so as mobile phone) this
results looks satisfying.

Signed-off-by: Karol Duda <karolwojciechduda@gmail.com>

### CODE REVIEW

File: drivers/sensor/ms5611/Kconfig [hunk 1]
```diff
@@ -0,0 +1,11 @@
+# MS5611 barometer
+
+# Copyright (c) 2022, Karol Duda
+# SPDX-License-Identifier: Apache-2.0
+
+menuconfig MS5611
+    bool "MS5611 Barometer"
+    default y
+    depends on I2C
```
Comment (galak) [inline drivers/sensor/ms5611/Kconfig:9]: add
```
depends on DT_HAS_MEAS_MS5611_ENABLED
```

Change `depends on I2C` to `select I2C`

**Review by galak — DISMISSED** [review]
Some Kconfig changes

**Comment by galak** [comment]
Add to `tests/drivers/build_all/sensor/i2c.dtsi`

File: drivers/sensor/ms5611/ms5611.c [hunk 2]
```diff
@@ -0,0 +1,551 @@
...
+#define DT_DRV_COMPAT meas_ms5611
+
+#include <zephyr/zephyr.h>
+#include <zephyr/drivers/sensor.h>
+#include <zephyr/device.h>
+#include <zephyr/drivers/i2c.h>
+#include <zephyr/sys/byteorder.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(ms5611);
```
Comment (MaureenHelm) [inline drivers/sensor/ms5611/ms5611.c:16]:
```suggestion
LOG_MODULE_REGISTER(ms5611, CONFIG_SENSOR_LOG_LEVEL);
```

File: drivers/sensor/ms5611/ms5611.c [hunk 3]
```diff
@@ -0,0 +1,551 @@
...
+		}
+
+		/* convert data to little-endian */
+		data->prom[i] = sys_get_be16(data_bytes);
+	}
+
+	return 0;
+}
+
+static uint8_t ms5611_calculate_crc4(const struct device *dev)
```
Comment (MaureenHelm) [inline drivers/sensor/ms5611/ms5611.c:406]: Can you use a builtin CRC implementation? `include/zephyr/sys/crc.h`

Comment (karold2412) [reply drivers/sensor/ms5611/ms5611.c:406]: Actually, I wanted to do so... Please correct me if I am wrong, but I see in `include/zephyr/sys/crc.h` only functions for 32, 16, 8 and 7 bit CRC calculation. In this sensor datasheet it is stated that we need 4 bit CRC. I didn't found any information how to do that using built in funtions. Is there some other way?

Comment (catch-twenty-two) [reply drivers/sensor/ms5611/ms5611.c:406]: if you don't see one you could add to it, unless it's specific to the ms5611 and nothing else

Comment (karold2412) [reply drivers/sensor/ms5611/ms5611.c:406]: According to this application note:
https://www.te.com/commerce/DocumentDelivery/DDEController?Action=showdoc&DocId=Specification+Or+Standard%7FAN520_C-code_example_for_MS56xx%7FA%7Fpdf%7FEnglish%7FENG_SS_AN520_C-code_example_for_MS56xx_A.pdf%7FCAT-BLPS0003
I am afraid that it is custom solution.

File: drivers/sensor/ms5611/ms5611.c [hunk 4]
```diff
@@ -0,0 +1,551 @@
...
+	/* Request temperature conversion */
+	i2c_cmd = data->osr[MS5611_OSR_TEMP_IDX].read_cmd;
+	ret = i2c_write_dt(&cfg->i2c_bus, &i2c_cmd, sizeof(i2c_cmd));
+	if (ret < 0) {
+		LOG_ERR("Failed to request temperature conversion");
+		return ret;
+	}
+
+	/* Wait for conversion to be finished */
+	k_sleep(K_USEC(data->osr[MS5611_OSR_TEMP_IDX].resp_time));
```
Comment (catch-twenty-two) [inline drivers/sensor/ms5611/ms5611.c:234]: A couple of questions, what is the longest time period that this, combined with the temp reading for calibration could cause blocking on the calling thread? Are there any status bits/interrupt lines to guarantee you have a new reading?

Comment (karold2412) [reply drivers/sensor/ms5611/ms5611.c:234]: According to the datasheet: values provided in lines 49-54 of ms5611.c file are "Max" conversion values. Thus, I suppose these are the worst case scenario values.
Unfortunately there is no interrupt line and the only way to get the information "if conversion has completed" is to ask the chip in next command sequence. If something went wrong (for example conversion wasn't completed) then the result is 0. This is why I am checking if the result is different from 0 in lines 251 and 282.

**Comment by catch-twenty-two** [comment]
lgtm nice job

**Review by catch-twenty-two — DISMISSED** [review]
lgtm nice job

File: drivers/sensor/ms5611/ms5611.c [hunk 5]
```diff
@@ -0,0 +1,551 @@
...
+	/* Second order temperature compensation */
+	if (temp < 2000) {
+		t2 = ((dT ^ 2) >> 31);
+		off2 = 5 * ((temp - 2000) ^ 2) / 2;
+		sens2 = 5 * (temp - 2000) ^ 2 / 4;
+
+		if (temp < -1500) {
+			off2 = off2 + (7 * (temp + 1500) ^ 2);
+			sens2 = sens2 + (11 * ((temp + 1500) ^ 2)) / 2;
+		}
```
Comment (MaureenHelm) [inline drivers/sensor/ms5611/ms5611.c:300]: Please document the source of this temperature compensation algorithm. Link to a datasheet and section number, for example.

File: drivers/sensor/ms5611/ms5611.c (hunk 3 above)
Comment (MaureenHelm) [reply drivers/sensor/ms5611/ms5611.c:406]: Please add some comments to the code or commit message explaining why you've implemented a custom CRC

**Review by MaureenHelm — CHANGES_REQUESTED** [review]

File: drivers/sensor/ms5611/ms5611.c (hunk 5 above)
Comment (karold2412) [reply drivers/sensor/ms5611/ms5611.c:311]: Added few lines above

File: drivers/sensor/ms5611/ms5611.c (hunk 3 above)
Comment (karold2412) [reply drivers/sensor/ms5611/ms5611.c:417]: Added in the commit message :+1:

**Comment by karold2412** [comment]
@MaureenHelm - I applied changes you requested. Could you re- review?

**Pull request closed without merging**